src/Clash/Cores/Etherbone/Base.hs

```diff
@@ -0,0 +1,291 @@
+--
+-- By using a bypass line the state machine of the WishboneMaster can be made
+-- simpler and the RecordBuilder can know about the number of reads and writes
+-- before an operation is finished.
+data Bypass addrWidth = Bypass
+  -- | The 'RecordHeader' from the @_meta@ field.
+  { _bpHeader :: RecordHeader
+  -- | The @BaseRetAddr@ that needs to be returned if a read operation has is
+  -- sent.
+  , _bpBase   :: Maybe (BitVector addrWidth)
+  -- | The @_abort@ signal from the PacketStream. If this is @True@, no new data
+  -- will come from 'wishboneMasterC' and 'recordBuilderC' should be set back
+  -- into its initial state.
+  , _bpAbort  :: Bool
+  } deriving (Generic, NFDataX, NFData, Show, ShowX, Eq)
+
+type ByteSize dat = DivRU (BitSize dat) 8
+
+-- | Extract 'EBHeader' data from a @PacketStream@ into the metadata.
+-- With a 64-bit bus this shifts the data-stream by 4 bytes.
+etherboneDepacketizerC :: forall dom dataWidth .
+ ( HiddenClockResetEnable dom, KnownNat dataWidth, 4 <= dataWidth )
+ => Circuit (PacketStream dom dataWidth ()) (PacketStream dom dataWidth EBHeader)
+etherboneDepacketizerC = depacketizerC @_ metaMap
+  where
+    metaMap :: EBHeader -> () -> EBHeader
+    metaMap hdr _ = hdr
+
+-- | Concat 'EBHeader' metadata back into the data stream. This gives
+-- backpressure for one cycle.
+etherbonePacketizerC :: forall dom dataWidth .
+  ( HiddenClockResetEnable dom, KnownNat dataWidth, 4 <= dataWidth )
+  => Circuit (PacketStream dom dataWidth EBHeader) (PacketStream dom dataWidth ())
+etherbonePacketizerC = packetizerC (const ()) metaMap
+  where
+    metaMap :: EBHeader -> EBHeader
+    metaMap hdr = hdr
+
+-- | Strips 4 bytes from the start of a packet in the case of a 64-bit wishbone
+-- bus and passes the input for a 32-bit bus.
+-- This is needed to get rid of the 'potential padding' in the case of a 64-bit
+-- bus.
+etherbonePaddingStripperC :: forall dom dataWidth .
+  ( HiddenClockResetEnable dom
+  , KnownNat dataWidth
+  , 4 <= dataWidth
+  )
+  => Circuit (PacketStream dom dataWidth EBHeader) (PacketStream dom dataWidth EBHeader)
+etherbonePaddingStripperC = case compareSNat (SNat @dataWidth) d4 of
+  SNatLE -> idC
+  SNatGT -> depacketizerC metaMap
+  where
+    metaMap :: Vec (dataWidth-4) (BitVector 8) -> EBHeader -> EBHeader
+    metaMap _ hdr = hdr
+
+-- | Adds 4 bytes of padding to the start of a packet in the case of a 64-bit
+-- wishbone bus. It passes the input for a 32-bit bus.
+-- Needed to apply the 'potential padding' in the case of a 64-bit wishbone bus.
+etherbonePaddingAdderC :: forall dom dataWidth .
+  ( HiddenClockResetEnable dom
+  , KnownNat dataWidth
+  , 4 <= dataWidth
+  )
+  => Circuit (PacketStream dom dataWidth EBHeader) (PacketStream dom dataWidth EBHeader)
+etherbonePaddingAdderC = case compareSNat (SNat @dataWidth) d4 of
+  SNatLE -> idC
+  SNatGT -> packetizerC id padMap
+  where
+    padMap :: EBHeader -> Vec (dataWidth-4) (BitVector 8)
+    padMap _ = repeat 0
```

Comment:
Same but even more type application:
```
  SNatGT -> packetizerC @_ @_ @_ @_ @(Vec (dataWidth-4) (BitVector 8)) id (const (repeat 0))
```

Reply:
While this can once again be solved through the `headerBytes` tyarg, it is pretty unreadable because you have to count tyargs to see what was applied:
```haskell
packetizerC @_ @_ @_ @_ @_ @(dataWidth-4) id (const (repeat 0))
```

In this case, I prefer the original code. It's much clearer what it does.